saleor/graphql/attribute/mutations.py

```diff
@@ -747,14 +748,16 @@ def success_response(cls, instance):
 
     @classmethod
     def post_save_action(cls, info, instance, cleaned_input):
-        variants = product_models.ProductVariant.objects.filter(
-            Exists(instance.variantassignments.filter(variant_id=OuterRef("id")))
-        )
-
-        product_models.Product.objects.filter(
-            Q(Exists(instance.productassignments.filter(product_id=OuterRef("id"))))
-            | Q(Exists(variants.filter(product_id=OuterRef("id"))))
-        ).update(search_index_dirty=True)
+        with transaction.atomic():
+            variants = product_models.ProductVariant.objects.filter(
+                Exists(instance.variantassignments.filter(variant_id=OuterRef("id")))
+            )
+            qs = product_models.Product.objects.select_for_update(of=("self",)).filter(
+                Q(Exists(instance.productassignments.filter(product_id=OuterRef("id"))))
+                | Q(Exists(variants.filter(product_id=OuterRef("id"))))
+            )
+            list(qs)  # evaluate qs to enforce acquiring the lock.
```

Comment:
What is the benefit of locking rows for an update immediately before the update? Update and select-for-update both use the same type of lock. This line generates an extra query but does not change how the code works.

Reply:
without this line, no SELECT FOR UPDATE statement is being executed (only the UPDATE WHERE is executed) Therefore no lock seems to get acquired and in high-concurrency situation it deadlocks.

Reply:
`SELECT … FOR UPDATE` and `UPDATE` both create the same type of lock.